packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrix.hpp

```diff
@@ -70,270 +72,345 @@ class Distributor;
+    // object shows an error and if its associated `bad_index` is less than the
+    // dst.bad_index (if dst shows errors).
+    LO invalid = Tpetra::Details::OrdinalTraits<LO>::invalid();
+    if (src.bad_index != invalid) {
+      // An error in the src, check if whether:
+      //   1. The dst shows errors
+      //   2. If dst does show errors, if src bad_index is less than its bad
+      //      index
+      if (dst.bad_index == invalid || src.bad_index < dst.bad_index) {
+        dst.bad_index = src.bad_index;
+        dst.num_ent = src.num_ent;
+        dst.offset = src.offset;
+        dst.expected_num_bytes = src.expected_num_bytes;
+        dst.num_bytes = src.num_bytes;
+        dst.buf_size = src.buf_size;
+        dst.wrong_num_bytes_error = src.wrong_num_bytes_error;
+        dst.out_of_bounds_error = src.out_of_bounds_error;
+      }
+    }
   }
 
-  return numValid;
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const LO i, value_type& dst) const
+  {
+    const LO local_row = import_lids_[i];
+    const size_t num_bytes = num_packets_per_lid_(i);
 
-}
+    if (num_bytes > 0) { // there is actually something in the row
 
-template<class LocalMatrix, class LocalMap>
-bool
-unpackCrsMatrixRow(LocalMatrix& lclMatrix,
-                   LocalMap& lclColMap,
-                   std::unique_ptr<std::string>& errStr,
-                   typename LocalMatrix::value_type* const valInTmp,
-                   typename LocalMap::global_ordinal_type* const indInTmp,
-                   const size_t tmpSize,
-                   const char* const valIn,
-                   const char* const indIn,
-                   const size_t numEnt,
-                   const typename LocalMatrix::ordinal_type lclRow,
-                   const Tpetra::CombineMode combineMode,
-                   const bool atomic)
-{
-  if (tmpSize < numEnt || (numEnt != 0 && (valInTmp == NULL || indInTmp == NULL))) {
-    return false;
+      const size_t buf_size = imports_.size();
+      const char* const num_ent_beg = imports_.ptr_on_device() + offsets_(i);
+      const char* const num_ent_end = num_ent_beg + sizeof(LO);
+
+      // Now we know how many entries to expect in the received data
+      // for this row.
+      LO num_ent = Teuchos::ScalarTraits<LO>::zero();
+      memcpy(&num_ent, num_ent_beg, sizeof(LO));
+
+      const char* const val_beg = num_ent_end;
+      const char* const val_end = val_beg + static_cast<size_t>(num_ent) * sizeof(IST);
+      const char* const ind_beg = val_end;
+      const size_t expected_num_bytes = sizeof(LO) +
+        static_cast<size_t>(num_ent) * (sizeof(IST) + sizeof(GO));
+
+      if (expected_num_bytes > num_bytes) {
+        dst.wrong_num_bytes_error = true;
+      }
+      else if (offsets_(i) > buf_size || offsets_(i) + num_bytes > buf_size) {
+        dst.out_of_bounds_error = true;
+      }
+      else {
+
+        // FIXME (mfh 23 Mar 2017) CrsMatrix_NonlocalSumInto_Ignore test
```

Comment:
Are these comments still valid?  If so, feel free to relabel.  If not, delete them.